### DISCUSSION TITLE
[Share] Auto-run Better Notes template when opening an item

### DISCUSSION THREAD

**Discussion by windingwind** [windingwind/zotero-actions-tags#108] [Action Scripts]
## Description

Auto-run Better Notes template when opening an item

- During the first run (opening file), you will be asked to choose a template to be executed. Your choice will be remembered.
- When opening an item, create an auto-template note and insert the chosen template to the note using the opened item as template input (see [about-note-template#item](https://github.com/windingwind/zotero-better-notes/blob/master/docs/about-note-template.md#item) for more details about BN template). If an auto-template is already generated, skip it.

> Do not run builtin note templates of the BN plugin

> If you want to reset the remembered template, go to preferences - advanced - config editor - search `extensions.actionsTags.customAction.autoRunBNTemplate` and reset.

> Require [Better Notes for Zotero](https://github.com/windingwind/zotero-better-notes) >= 1.1.4-21

<img width="1108" alt="image" src="https://github.com/windingwind/zotero-tag/assets/33902321/547af299-5ea8-4c41-aea1-3c6dc52e1a7e">

## Action Script

Event: Open File
Operation: Script
Data:

```js
// Auto-run Better Notes template when opening an item for the first time
// @author windingwind
// @link https://github.com/windingwind/zotero-tag/discussions/108
// @usage Event=Open File

const window = require("window");

if (!Zotero.BetterNotes?.api.note?.insert) {
	return "[Action:Auto-Template] Better Notes for Zotero > 1.1.4-21 is not installed or disabled.";
}
if (!item) {
	return "[Action:Auto-Template] Target item is empty";
}

// Check if an auto-template note already exists
const autoTemplateTag = "auto-template";
const notes = Zotero.Items.get(item.getNotes());
if (notes.some(note => note.getTags().some(obj => obj.tag === autoTemplateTag))) {
	return "[Action:Auto-Template] Auto-template already exists";
}

const prefsKey = "extensions.actionsTags.customAction.autoRunBNTemplate";

// Get template name
let templateName = Zotero.Prefs.get(prefsKey, true);
let templateContent = Zotero.BetterNotes.api.template.getTemplateText(templateName);
if (!templateContent) {
	// Template name is not defined or invalid. Get from input
	const templateNames = Zotero.BetterNotes.api.template.getTemplateKeys();

	function selectItems(itemList) {
		var io = { dataIn: itemList, dataOut: null };
		window.openDialog("chrome://scaffold/content/select.xhtml",
			"_blank", "chrome,modal,centerscreen,resizable=yes", io);
		return io.dataOut;
	}

	templateName = Object.values(selectItems(templateNames))[0];
	templateContent = Zotero.BetterNotes.api.template.getTemplateText(templateName);
	if (!templateContent) {
		return "[Action:Auto-Template] Template is invalid";
	}
	Zotero.Prefs.set(prefsKey, templateName, true);
}

const parentItem = Zotero.Items.getTopLevel([item])[0];

const noteItem = new Zotero.Item("note");
noteItem.libraryID = parentItem.libraryID;
noteItem.parentID = parentItem.id;
noteItem.addTag(autoTemplateTag, 0);
await noteItem.saveTx();

let html = "";
if (templateName.toLowerCase().startsWith("[item]")) {
	html = await Zotero.BetterNotes.api.template.runItemTemplate(templateName, {
		itemIds: [parentItem.id],
		targetNoteId: noteItem.id,
	});
} else {
	html = await Zotero.BetterNotes.api.template.runTextTemplate(templateName, {
		targetNoteId: noteItem.id,
	});
}
await Zotero.BetterNotes.api.note.insert(
	noteItem,
	html,
	-1,
);

return `[Action:Auto-Template] Note ${noteItem.getNoteTitle()} is created on item ${item.getField("title")} from template ${templateName}`;
```

**Comment by wklimowicz** [comment]
This will create a new note on every object that doesn't have an auto-template note. If users want to change this to run only if there is no other note already attached to an item, change the code starting with `// Check if an auto-template note already exists`, to:

```js
// Check if any note exists for the item
const autoTemplateTag = "auto-template";
const notes = Zotero.Items.get(item.getNotes());
if (notes.length > 0) {
    return "A note exists for this item";
}
```

**Reply by windingwind** [reply]
Better start a new discussion thread and share there, as this is a different action.

**Reply by wklimowicz** [reply]
Added as #109.

**Comment by niao004** [comment]


**Reply by windingwind** [reply]
the image link is broken

**Comment by janeluyip** [comment]
I do not find the extensions.actionsTags.customAction.autoRunBNTemplate
which extensions it belongs to？

**Reply by windingwind** [reply]
i use the prefix of the actions & tags plugin, but it does not belong to any plugin.

**Reply by janeluyip** [reply]
期待早日推出正式版

**Comment by horaynia** [comment]
大佬，我添加后运行后的模版和您图里的模版不一样呢？请问是咋回事儿
<img width="344" alt="image" src="https://github.com/windingwind/zotero-actions-tags/assets/50537627/cfa905dd-a973-4ca1-8367-cfae8e39950d">

**Reply by windingwind** [reply]
取决于你设置的模板是什么。本动作只是调用特定模板

**Reply by horaynia** [reply]
大佬，我有和您图示里类似的模版，但是复制进去后运行错误。不知道是咋回事儿？请问您说的设置的模版是啥意思？

**Reply by horaynia** [reply]
想用这个模版：https://github.com/windingwind/zotero-better-notes/discussions/771#discussioncomment-7432015

**Reply by windingwind** [reply]
1. 不借助actions插件是否可以正常使用模板
2. 运行错误指的是什么

**Reply by horaynia** [reply]
不借助action，可以单独使用该模版（https://github.com/windingwind/zotero-better-notes/discussions/771#discussioncomment-7432015）
将模版填入后，提示如下图错误
<img width="302" alt="image" src="https://github.com/windingwind/zotero-actions-tags/assets/50537627/a5bf214b-9806-4e47-a349-dc8bfd75eb4a">

**Reply by windingwind** [reply]
你不会是把BN模板复制到了action的动作脚本里了吧

**Reply by horaynia** [reply]
我导入您的上面的代码，运行后笔记模版不是您图示的那个。是我之前构建的那个模版。就很奇怪。但是我已经将它禁用了啊。我删除它看看

**Reply by horaynia** [reply]
难道是和bn的默认模版有关？

**Reply by windingwind** [reply]
在第一次运行动作时弹窗选择需要的笔记模板。如果已经运行过一次，要换成其他笔记模板，则需要通过高级设置修改`extensions.actionsTags.customAction.autoRunBNTemplate`为目标笔记模板的名称。如何修改zotero高级设置请善用搜索

**Reply by horaynia** [reply]
弄好了，感谢大佬

**Reply by horaynia** [reply]
> 在第一次运行动作时弹窗选择需要的笔记模板。如果已经运行过一次，要换成其他笔记模板，则需要通过高级设置修改`extensions.actionsTags.customAction.autoRunBNTemplate`为目标笔记模板的名称。如何修改zotero高级设置请善用搜索

感谢大佬

**Reply by iampearl918** [reply]
请问：
本帖子给出的是action的代码，按我目前的理解，是在Zotero Actions & Tags插件中要进行的设置。

>During the first run (opening file), you will be asked to choose a template to be executed. 

图片中笔记模版的代码是否应该要另外下载呢？这个模版的代码是哪个呢？

**Reply by windingwind** [reply]
关于其他插件的使用方法，参见对应插件文档

**Reply by iampearl918** [reply]
好的，已经找到了类似模版，要图片中的效果需要稍作改写

**Reply by horaynia** [reply]
> > 在第一次运行动作时弹窗选择需要的笔记模板。如果已经运行过一次，要换成其他笔记模板，则需要通过高级设置修改`extensions.actionsTags.customAction.autoRunBNTemplate`为目标笔记模板的名称。如何修改zotero高级设置请善用搜索
> 
> 感谢大佬

大佬,这个打开后没有找到修改默认笔记模版的地方.请问在哪儿修改默认模版呢
![Uploading HapiGo_2024-11-10_19.02.48.png…]()

**Comment by iampearl918** [comment]
“During the first run (opening file), you will be asked to choose a template to be executed.”为什么我打开笔记工作区，没有弹出来呢？
而且若点击“新建工作区笔记”，弹出提示“JavaScript Application-请选择一个有效的分类“，只给出”Ok“选项，点按之后就没有任何反应了，创建不了新笔记。

**Reply by windingwind** [reply]
> “During the first run (opening file), you will be asked to choose a template to be executed.”为什么我打开笔记工作区，没有弹出来呢？

first run=首次运行本action，与工作区无关

> 而且若点击“新建工作区笔记”，弹出提示“JavaScript Application-请选择一个有效的分类“，只给出”Ok“选项，点按之后就没有任何反应了，创建不了新笔记。

字面意思，当前选中的分类不合理（例如选中了feed或回收站等，无法在此处创建笔记）。需在library的左侧栏选择合理的collection

**Reply by iampearl918** [reply]
> first run=首次运行本action，与工作区无关

谢谢您的耐心解释，是我将基础的概念搞混了，因为首次接触和摸索Markdown和BN这类笔记软件，突然接触到大量新概念和功能，有些蒙没弄明白。还有个问题想请教：

在查看BN相关资料时，经常出现的“item”，是指Zotero里存储的单篇文章吗？
这样的理解，让我在很多地方能暂时读的通。
有的地方，我又感觉item是指单篇文献中的“元素”，也就是图表、标题、期刊名称、作者等等。
但又似乎都不对，比如您这个帖子正文中链接里[about-note-template#item](https://github.com/windingwind/zotero-better-notes/blob/master/docs/about-note-template.md#item)，我实在弄不明白item是指什么。

  因为BN相关使用说明基本是英文的，对于我都是一些新的计算机软件术语，看得懂英文单词、却不太明白指什么。谢谢您
@windingwind

**Reply by windingwind** [reply]
https://zotero-chinese.com/user-guide/add-items

善用搜索

**Comment by fredericky123** [comment]
这里能不能完善一下，事件为关闭tag时，先删除已有的annotation，再自动生成一个，因为每次打开tag的目的是阅读更新，阅读完标注后期待模板能自动更新一下，就免去手动导入了

**Reply by DWCY3574** [reply]
找到好方法了吗？楼下的大佬分享的是好用，但自己编辑的笔记更新后就没了

**Comment by thebetter2** [comment]
根据颜色分类整理注释到笔记中的笔记模板需要每次打开时更新笔记。因此修改一下大佬的action 插件内容为 检测到存在 auto-template标签笔记则删除它，然后继续创建。
action插件配置如图
![image](https://github.com/user-attachments/assets/4b896693-70f7-4d3e-81e1-f46fedb72168)

action代码如下。另附传送门https://github.com/windingwind/zotero-better-notes/discussions/668#discussion-5579097   可以根据注释颜色分类整理进笔记的better note笔记模板。二者可以结合使用。
```
// Auto-run Better Notes template when opening an item for the first time
// @author windingwind
// @link https://github.com/windingwind/zotero-tag/discussions/108
// @usage Event=Open File
const Zotero = require("Zotero");
const window = require("window");

if (!Zotero.BetterNotes?.api.note?.insert) {
    return "[Action:Auto-Template] Better Notes for Zotero > 1.1.4-21 is not installed or disabled.";
}
if (!item) {
    return "[Action:Auto-Template] Target item is empty";
}

// Check if an auto-template note already exists and delete it
const autoTemplateTag = "auto-template";
const notes = Zotero.Items.get(item.getNotes());

for (const note of notes) {
    if (note.getTags().some(obj => obj.tag === autoTemplateTag)) {
        // Delete the existing note with the "auto-template" tag
        await Zotero.Items.erase(note.id);
    }
}

const prefsKey = "extensions.actionsTags.customAction.autoRunBNTemplate";

// Get template name
let templateName = Zotero.Prefs.get(prefsKey, true);
let templateContent = Zotero.BetterNotes.api.template.getTemplateText(templateName);
if (!templateContent) {
    // Template name is not defined or invalid. Get from input
    const templateNames = Zotero.BetterNotes.api.template.getTemplateKeys();

    function selectItems(itemList) {
        var io = { dataIn: itemList, dataOut: null };
        window.openDialog("chrome://zotero/content/ingester/selectitems.xhtml",
            "_blank", "chrome,modal,centerscreen,resizable=yes", io);
        return io.dataOut;
    }

    templateName = Object.values(selectItems(templateNames))[0];
    templateContent = Zotero.BetterNotes.api.template.getTemplateText(templateName);
    if (!templateContent) {
        return "[Action:Auto-Template] Template is invalid";
    }
    Zotero.Prefs.set(prefsKey, templateName, true);
}

const parentItem = Zotero.Items.getTopLevel([item])[0];

const noteItem = new Zotero.Item("note");
noteItem.libraryID = parentItem.libraryID;
noteItem.parentID = parentItem.id;
noteItem.addTag(autoTemplateTag, 0);
await noteItem.saveTx();

let html = "";
if (templateName.toLowerCase().startsWith("[item]")) {
    html = await Zotero.BetterNotes.api.template.runItemTemplate(templateName, {
        itemIds: [parentItem.id],
        targetNoteId: noteItem.id,
    });
} else {
    html = await Zotero.BetterNotes.api.template.runTextTemplate(templateName, {
        targetNoteId: noteItem.id,
    });
}
await Zotero.BetterNotes.api.note.insert(
    noteItem,
    html,
    -1,
);

return `[Action:Auto-Template] Note ${noteItem.getNoteTitle()} is created on item ${item.getField("title")} from template ${templateName}`;

```

**Reply by DWCY3574** [reply]
使用这个自己输入的笔记就没了

**Reply by fredericky123** [reply]
所以我目前想的用另一个笔记保存自己输入的笔记，如果有更好的合二为一的办法，期待了

**Reply by kepbod** [reply]
如何和颜色标签模版联合使用？我看是要有auto-template的 tag的 note？https://github.com/windingwind/zotero-better-notes/discussions/668#discussion-5579097

**Reply by thebetter2** [reply]
> 如何和颜色标签模版联合使用？我看是要有auto-template的 tag的 note？[windingwind/zotero-better-notes#668 (comment)](https://github.com/windingwind/zotero-better-notes/discussions/668#discussion-5579097)
tag不用手动写，是自动创建的。
使用方法很简单，你首次运行这个action的时候会出现如下图，选择模板界面，之后每次打开文献就会自动按照这个模板生成。
![image](https://github.com/user-attachments/assets/5942b648-026b-4e92-b29b-bfb1c058f44e)
打开生成的笔记后，就会在标签页看的生成的标签。你的其他用别的模板生成的、没有这个标签的笔记不会被删除。
![image](https://github.com/user-attachments/assets/e0c8210e-efa4-4f9c-8dc5-0a23613b04df)
如果相更换或者同时自动生成多个别的笔记模板，需要删除这个action，新建或者复制一个新的，选别的模板就行，需要设置时，将标签改成不同的就行。

**Reply by wzr-ms** [reply]
提出一个问题，更新之后，我在笔记上写的批注也被删除了

**Comment by xjtusjtu** [comment]
如果模板数量足够多，会导致模板显示不完全，以及 SelectAll，Deselect All，OK，Cancel 等按钮不显示，请问此时应该如何修改 window.openDialog 代码？

例如
```js
let tags = await Zotero.Tags.getAll();
let templateNames = tags.map(t =>t.tag);

function selectItems(itemList) {
        var io = { dataIn: itemList, dataOut: null };
        window.openDialog("chrome://zotero/content/ingester/selectitems.xhtml",
            "_blank", "chrome,modal,centerscreen,resizable=yes", io);
        return io.dataOut;
}

templateName = Object.values(selectItems(templateNames.slice(0,40)));
console.log(templateName);
```

![image](https://github.com/user-attachments/assets/4b6aa004-e535-4044-aa8c-b1ef3ff59644)

**Comment by ramonmi** [comment]
- Zotero 7.0.9及以上版本，将select.xhtml由chrome://zotero/content/ingester/selectitems.xhtml移至chrome://scaffold/content/select.xhtml，see https://github.com/zotero/zotero/pull/4747/files#diff-a5cc4a0d67b871fc25c8a4710168be2043bfdc47b1b43eecb76b6fe8ea97db80R1200
    - 因此上述代码需做如下更改
```js
window.openDialog("chrome://zotero/content/ingester/selectitems.xhtml",
```
修改为
```js
// window.openDialog("chrome://zotero/content/ingester/selectitems.xhtml",
window.openDialog("chrome://scaffold/content/select.xhtml",
```
即可正常使用，新版本增加了滑动翻页显示，避免模板较多时显示不全。

- BTW，better notes更新了新的对话框选择（如下图按钮，现在是新的重构对话框），请问 @windingwind 可以使用脚本调用吗？谢谢！
![屏幕截图 2024-10-29 104742](https://github.com/user-attachments/assets/f2f85a9c-9cf6-4d08-b75e-769c0012ec62)

**Reply by xjtusjtu** [reply]
感谢！

**Reply by hogan-github** [reply]
太牛了

**Reply by Dani-v-tue** [reply]
Hey, I used google translate to join this conversation so I'm not sure if I missed something, but:

```
window.openDialog("chrome://scaffold/content/select.xhtml",
```

Broke the popup on my machine (MacOS), and using
```
window.openDialog("chrome://zotero/content/ingester/selectitems.xhtml",
```
 fixed it. Also relevant for anyone else that sees this :).

**Comment by Calorion** [comment]
I installed this, and it worked, but it automatically chose the wrong template; it never asked what template I wanted. 

I tried resetting `extensions.actionsTags.customAction.autoRunBNTemplate`, but…what does “reset” mean? I changed it from True to False, and it stopped working entirely. I changed it back to True, and it still doesn’t function.

**Reply by windingwind** [reply]
The template is updated, please clear that field and try again.

**Reply by Calorion** [reply]
That’s working, but now it’s bringing up the dialog asking for which template to use every time.

**Reply by Calorion** [reply]
Is this a bug, or am I doing something wrong?

**Reply by ramonmi** [reply]
> but now it’s bringing up the dialog asking for which template to use every time.

Based on my test, the template works well. May be you can check the `extensions.actionsTags.customAction.autoRunBNTemplate` argument in the Zotero-Settings-Advanced-Config Editor, normally it is a string titled `[Item] template name`.

**Reply by Calorion** [reply]
Deleting that argument (and changing it from Boolean to String, not sure that mattered) worked, thanks.

**Comment by JiQ6357** [comment]
你好 @windingwind，感谢分享这个自动运行 Better Notes 模板的示例！

我在整理一篇 Zotero + Codex 文献阅读教程，之前参考过这里的接口用法，在 AI 辅助下做了一个适合自己流程的脚本：固定调用笔记模板，父条目已有笔记就跳过，并加了模板检查和错误提示。

想确认一下：下面这份脚本中，涉及你示例的部分，是否允许在保留你的署名和本帖链接的情况下，随教程以 MIT 许可公开分享？如果需要遵循其他许可，也请告诉我。谢谢！

<details>
<summary>当前使用的 AutoGenerateNote.js（点击展开）</summary>

```javascript
// 打开 PDF 时：
// 如果文献主条目没有任何子笔记，则使用 Better Notes 模板创建一条笔记；
// 如果已经存在任意笔记，则跳过。

const expectedTemplateName = "[item]AI简略文献笔记V1";

try {
    // 1. 获取当前打开的 PDF 附件
    if (!item) {
        return "[AutoGenerateNote] 未获取到当前打开的附件";
    }

    // 2. 获取附件所属的文献主条目
    const parentItem = Zotero.Items.getTopLevel([item])[0];

    if (!parentItem) {
        return "[AutoGenerateNote] 无法获取文献主条目";
    }

    // 3. 只要主条目下已有任意笔记，就不再创建
    const existingNoteIDs = parentItem.getNotes();

    if (existingNoteIDs.length > 0) {
        return `[AutoGenerateNote] 已有 ${existingNoteIDs.length} 条笔记，跳过创建：${
            parentItem.getField("title")
        }`;
    }

    // 4. 获取 Better Notes API
    const bn = Zotero.BetterNotes?.api;

    if (
        !bn?.template?.getTemplateKeys ||
        !bn?.template?.getTemplateText ||
        !bn?.template?.runItemTemplate ||
        !bn?.note?.insert
    ) {
        return "[AutoGenerateNote] Better Notes API 不可用，请检查插件是否启用";
    }

    // 5. 从 Better Notes 实际模板列表中匹配模板
    const templateKeys = bn.template.getTemplateKeys();

    const templateName = templateKeys.find(name =>
        name.trim().toLowerCase() ===
        expectedTemplateName.trim().toLowerCase()
    );

    if (!templateName) {
        return `[AutoGenerateNote] 找不到模板：${expectedTemplateName}\n\n` +
            "Better Notes 当前识别到的模板：\n" +
            templateKeys
                .map(name => JSON.stringify(name))
                .join("\n");
    }

    // 6. 检查模板内容
    const templateContent =
        bn.template.getTemplateText(templateName);

    if (!templateContent) {
        return `[AutoGenerateNote] 模板内容为空或无法读取：${templateName}`;
    }

    // 7. 创建主条目下的子笔记
    const noteItem = new Zotero.Item("note");

    noteItem.libraryID = parentItem.libraryID;
    noteItem.parentID = parentItem.id;

    await noteItem.saveTx();

    // 8. 运行 Better Notes 模板
    const html = await bn.template.runItemTemplate(
        templateName,
        {
            itemIds: [parentItem.id],
            targetNoteId: noteItem.id,
        }
    );

    // 9. 将模板内容写入新建笔记
    await bn.note.insert(
        noteItem,
        html,
        -1
    );

    return `[AutoGenerateNote] 已创建笔记：${templateName}`;

} catch (error) {
    Zotero.logError(error);

    return `[AutoGenerateNote] 执行失败：${
        error?.message || String(error)
    }`;
}
```

</details>
